File: figcli2/cmd/callback.go
package cmd

import (
	"fmt"
	"io/ioutil"
	"os"

	"github.com/spf13/cobra"
)

const (
	versionNumber = "2"
)

func init() {
	callbackCmd.Flags().Bool("version", false, "Print version information")

	rootCmd.AddCommand(callbackCmd)
}

var callbackCmd = &cobra.Command{
	Use:    "callback",
	Hidden: true,
	Run: func(cmd *cobra.Command, args []string) {
		debug := os.Getenv("FIG_DEBUG") != ""

		if cmd.Flags().Lookup("version").Value.String() == "true" {
			fmt.Println(versionNumber)
			os.Exit(0)
		}

		if len(args) == 0 {
			if debug {
				fmt.Println("callback must include a handlerId")
			}
			os.Exit(1)
		}

		// Check if data is on stdin
		stdin := os.Stdin
		fi, err := stdin.Stat()
		if err != nil {
			panic(err)
		}
		size := fi.Size()
		if size == 0 {
			if debug {
				fmt.Println("No data on stdin!")
			}
			os.Exit(1)
		}

		// Get the handlerId and filename/exitcode if present
		handlerId := args[0]
		var filename string
		exitcode := "-1"

		if debug {
			fmt.Println("handlerId:", handlerId)
		}

		if len(args) == 3 {
			if debug {
				fmt.Printf("callback specified filepath (%s) and exitCode (%s) to output!\n", args[1], args[2])
			}

			filename = args[1]
			exitcode = args[2]
		} else {
			// Create tmp file
			file, err := ioutil.TempFile("/tmp/", "fig-callback-*")
			if err != nil {
				panic(err)
			}

			filename = file.Name()

			if debug {
				fmt.Printf("Created tmp file: %s\n", file.Name())
			}

			// Copy stdin to tmp file
			for {
				buf := make([]byte, 1024)
				n, err := stdin.Read(buf)
				if err != nil {
					panic(err)
				}

				// If we read 0 bytes, we're done
				if n == 0 {
					break
				}

				// Write to tmp file
				file.Write(buf[:n])

				if debug {
					fmt.Printf("Read %d bytes\n", size)
					fmt.Printf("%s\n", buf)
				}
			}

			if debug {
				fmt.Println("EOF!")
			}
		}

		// Send
		if debug {
			fmt.Println("Done reading from stdin!")
		}

		callback := fmt.Sprintf("fig pty:callback %s %s %s", handlerId, filename, exitcode)

		if debug {
			fmt.Printf("Sending '%s' over unix socket!\n", callback)
		}
	},
}

File: figterm/callback.c
#include "fig.h"
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <limits.h>

#define VERSION_NUMBER 2
#define MAX_BUFFER_SIZE 1024
#define MAX_HANDLER_ID_LEN 5
#define MAX_EXIT_CODE_LEN 3
int main(int argc, char *argv[]) {
  bool debug = (getenv("FIG_DEBUG") != NULL);

  if (argc < 2) {
    if (debug) printf("fig_callback must include a handlerId.\n");
    exit(1);
  }

  if (strcmp(argv[1], "-v") == 0 || strcmp(argv[1], "--version") == 0) {
    printf("%i\n", VERSION_NUMBER);
    exit(0);
  }

  // check if data on stdin
  if ((fseek(stdin, 0, SEEK_END), ftell(stdin)) > 0) {
    if (debug) printf("No data on stdin!\n");
    exit(1);
  }

  // get handler id from `fig-callback 6sD8n`
  char handlerId[MAX_HANDLER_ID_LEN + 1];
  memset(handlerId, '\0', sizeof(handlerId));
  strncpy(handlerId, argv[1], MAX_HANDLER_ID_LEN);
  if (debug) printf("handlerId: %s\n", handlerId);

  char filename[PATH_MAX];
  char exitcode[MAX_EXIT_CODE_LEN + 1] = "-1";
  if (argc == 4) {
    if (debug) printf("fig_callback specified filepath (%s) and exitCode (%s) to output!\n", argv[2], argv[3]);
    strncpy(filename, argv[2], PATH_MAX);
    
    memset(exitcode, '\0', sizeof(exitcode));
    strncpy(exitcode, argv[3], MAX_EXIT_CODE_LEN);
    goto send;
  }

  // todo(mschrage): determine exit code of previous command, if possible 

  // create tmp file
  char template[] = "/tmp/fig-callback-XXXXXX";

  strcpy(filename, template);

  int fd;
  fd = mkstemp(filename);
  FILE* fp = fdopen(fd, "w");

  if (debug) printf("Created tmp file: %s\n", filename);

  // read all of stdin
  char buffer[MAX_BUFFER_SIZE+1] = {0};
  size_t bytes;

  while ((bytes = fread(buffer, 1, MAX_BUFFER_SIZE, stdin)) >= 0) {
      if (debug) printf("Read %zu bytes\n", bytes);

      // write to file
      fwrite(buffer, sizeof(char), bytes, fp);
      if (debug) printf("%s\n", buffer);

      if (feof(stdin) || ferror(stdin)) { 
        fflush(fp);
        if (debug) printf("EOF!\n");
        break ;
      }
  }
  
  send:
  if (debug) printf("Done reading from stdin!\n");

  char *tmpbuf = malloc(strlen(filename) + sizeof(handlerId) + sizeof(exitcode) + sizeof(char) * 50);
  sprintf(
    tmpbuf,
    "fig pty:callback %s %s %s",
    handlerId,
    filename,
    exitcode
  );

  if (debug) printf("Sending '%s' over unix socket!\n", tmpbuf);

  // send to macOS app over unix socket
  fig_socket_send(strcat(tmpbuf, "\n"));
  free(tmpbuf);

}